src/main/java/com/google/api/generator/gapic/model/Message.java

```diff
@@ -80,6 +89,32 @@ public boolean hasResource() {
     return resource() != null;
   }
 
+  public void validateField(String fieldName, Map<String, Message> messageTypes) {
+    String[] subFields = fieldName.split("\\.");
+    Message nestedMessage = this;
+    for (int i = 0; i < subFields.length; i++) {
+      String subFieldName = subFields[i];
+      if (i < subFields.length - 1) {
+        Field field = nestedMessage.fieldMap().get(subFieldName);
+        nestedMessage = messageTypes.get(field.type().reference().fullName());
+        Preconditions.checkNotNull(
+            nestedMessage,
+            String.format(
+                MESSAGE_NOT_FOUND_ERROR_MESSAGE,
+                field.name(),
+                field.type().reference().simpleName()));
+      } else {
+        Preconditions.checkState(
+            !Strings.isNullOrEmpty(subFieldName),
+            String.format(EMPTY_FIELD_ERROR_MESSAGE, nestedMessage.name()));
+        Preconditions.checkState(
+            nestedMessage.fieldMap().containsKey(subFieldName),
+            String.format(FIELD_DOES_NOT_EXIST_ERROR_MESSAGE, nestedMessage.name(), subFieldName));
```

Comment:
I think it's better to restructure the loop to call `Field field = nestedMessage.fieldMap().get(subFieldName)` here right after this check, since the check goes hand in hand.

Actually, I wonder if it's better to remove the branching in the loop by only looping from 0 to subFields.length - 2 and then check the last element outside the loop. But I haven't tried.

Reply:
This validation logic is extracted from the existing code since the current implementation for implicit dynamic routing headers also need this validation, https://github.com/googleapis/gapic-generator-java/blob/b711799d7f631afd68f4a8f48b80b8ea5589da99/src/main/java/com/google/api/generator/gapic/protoparser/HttpRuleParser.java#L125. We can probably improve it if we want though.

Reply:
Restructured the validation logic. The pre-existing validation logic actually has a bug, it assumes the first field always exist, if it does not, there will be a null pointer exception.